File: api/js/src/validator/rules/empty-button-text.ts
import { createDiagnostic } from "../diagnostics.js";
import type { ValidationRule } from "../types.js";
import { RULE_IDS } from "./constants.js";
import { loc } from "./helpers.js";

export const emptyButtonTextRule: ValidationRule = {
  id: RULE_IDS.EMPTY_BUTTON_TEXT,
  description: "Button text must be non-empty",
  defaultSeverity: "error",
  run(ctx) {
    const diagnostics = [];
    for (const ref of ctx.normalized.flat) {
      const { button, rowIndex, columnIndex, flatIndex } = ref;
      const text = button.text;
      if (typeof text !== "string" || text.trim().length === 0) {
        diagnostics.push(
          createDiagnostic(
            RULE_IDS.EMPTY_BUTTON_TEXT,
            "Button text is empty or whitespace-only",
            "error",
            loc(rowIndex, columnIndex, flatIndex, "text"),
          ),
        );
      }
    }
    return diagnostics;
  },
};

File: api/python/telegram_inline_keyboard_builder/validator/rules/empty_button_text.py
from __future__ import annotations

from ..diagnostics import create_diagnostic
from ..types import Diagnostic, RuleContext, ValidationRule
from .constants import RULE_IDS
from .helpers import loc


def _run(ctx: RuleContext) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    for ref in ctx.normalized.flat:
        text = ref.button.get("text")
        if not isinstance(text, str) or text.strip() == "":
            diagnostics.append(
                create_diagnostic(
                    RULE_IDS.EMPTY_BUTTON_TEXT,
                    "Button text is empty or whitespace-only",
                    "error",
                    loc(ref.row_index, ref.column_index, ref.flat_index, "text"),
                )
            )
    return diagnostics


empty_button_text_rule = ValidationRule(
    id=RULE_IDS.EMPTY_BUTTON_TEXT,
    description="Button text must be non-empty",
    default_severity="error",
    run=_run,
)
